### PR TITLE
Add Last-Modified and If-Modified-Since headers support

### PR DESCRIPTION
**Pull request by Nyoroon**

### CODE REVIEW

File: prerender/app.py [hunk 1]
```diff
@@ -135,16 +137,32 @@ def _save_to_cache(key: str, data: bytes, format: str = 'html') -> None:
     if not skip_cache:
         try:
             data = await cache.get(url, format)
+            modified_since = cache.modified_since(url)
+            headers.update({'Last-Modified': formatdate(modified_since, usegmt=True)})
```
Comment (messense) [inline prerender/app.py]:
```python
headers['Last-Modified'] = formatdate(modified_since, usegmt=True)
```

File: prerender/app.py [hunk 2]
```diff
@@ -135,16 +137,32 @@ def _save_to_cache(key: str, data: bytes, format: str = 'html') -> None:
...
+                if_modified_since = 0
+
+            if modified_since and if_modified_since >= modified_since:
+                logger.info('Got 304 for %s in cache in %dms',
+                            url,
+                            int((time.time() - start_time) * 1000))
+                return response.text('', status=304, headers=headers)
+
             if data is not None:
+                headers.update({'X-Prerender-Cache': 'hit'})
```
Comment (messense) [inline prerender/app.py]: Same as above

File: prerender/app.py [hunk 3]
```diff
@@ -159,6 +177,8 @@ def _save_to_cache(key: str, data: bytes, format: str = 'html') -> None:
 
     try:
         data, status_code = await _render(request.app.prerender, url, format)
+        headers.update({'X-Prerender-Cache': 'miss'})
+        headers.update({'Last-Modified': formatdate(usegmt=True)})
```
Comment (messense) [inline prerender/app.py]:
```python
headers.update({'X-Prerender-Cache': 'miss', 'Last-Modified': formatdate(usegmt=True)})
```

File: prerender/cache/base.py [hunk 4]
```diff
@@ -2,8 +2,11 @@
...
 
 class CacheBackend:
-    async def get(self, key: str) -> Optional[bytes]:
+    async def get(self, key: str, format: str = 'html') -> Optional[bytes]:
         raise NotImplementedError
 
     def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -> None:
         raise NotImplementedError
+
+    def modified_since(self, key: str, format: str = 'html') -> int:
```
Comment (messense) [inline prerender/cache/base.py]: This method needs to be async or it might block the event loop.

File: prerender/cache/disk.py [hunk 5]
```diff
@@ -26,3 +27,7 @@ def __init__(self) -> None:
     def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -> None:
         compressed = lzma.compress(payload)
         self._cache.set(key + format, compressed, expire=ttl)
+
+    def modified_since(self, key: str, format: str = 'html') -> int:
```
Comment (messense) [inline prerender/cache/disk.py]: `async def`

File: prerender/cache/dummy.py [hunk 6]
```diff
@@ -8,3 +9,6 @@ class DummyCache(CacheBackend):
 
     def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -> None:
         pass
+
+    def modified_since(self, key: str, format: str = 'html') -> int:
```
Comment (messense) [inline prerender/cache/dummy.py]: `async def`

File: prerender/cache/s3.py [hunk 7]
```diff
@@ -59,9 +59,18 @@ def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -
             metadata={'url': key, 'ttl': ttl}
         )
 
+    def modified_since(self, key: str, format: str = 'html') -> int:
```
Comment (messense) [inline prerender/cache/s3.py]: `async def`

File: prerender/cache/s3.py [hunk 8]
```diff
@@ -59,9 +59,18 @@ def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -
...
+        path = self._filename(key, format)
+        try:
+            res = self.client.stat_object(S3_BUCKET, path)
```
Comment (messense) [inline prerender/cache/s3.py]: needs to run in executor like the `get` method.

**Review by messense — CHANGES_REQUESTED** [review]

**Review by Nyoroon — COMMENTED** [review]
fixed

File: prerender/cache/dummy.py [hunk 9]
```diff
@@ -8,3 +9,6 @@ class DummyCache(CacheBackend):
 
     def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -> None:
         pass
+
+    async def modified_since(self, key: str, format: str = 'html') -> int:
+        return time.time()
```
Comment (messense) [inline prerender/cache/dummy.py]: `int(time.time())` time.time() returns `float`

File: prerender/cache/s3.py [hunk 10]
```diff
@@ -59,9 +59,19 @@ def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -
             metadata={'url': key, 'ttl': ttl}
         )
 
+    async def modified_since(self, key: str, format: str = 'html') -> int:
+        path = self._filename(key, format)
+        loop = asyncio.get_event_loop()
+        try:
+            res = await loop.run_in_executor(None, self.client.stat_object, S3_BUCKET, path)
+        except (minio.error.NoSuchKey, asyncio.CancelledError):
+            return
```
Comment (messense) [inline prerender/cache/s3.py:68]: Needs to return a int value, maybe fallback to `int(time.time())`

Comment (Nyoroon) [reply prerender/cache/s3.py:68]: even `stat().st_time` returns float, maybe switch to float?

Comment (messense) [reply prerender/cache/s3.py:68]: as long as it works.

**Review by Nyoroon — COMMENTED** [review]
fixed return type of `modified_since()`

File: prerender/app.py [hunk 11]
```diff
@@ -135,16 +137,32 @@ def _save_to_cache(key: str, data: bytes, format: str = 'html') -> None:
     if not skip_cache:
         try:
             data = await cache.get(url, format)
+            modified_since = await cache.modified_since(url)
```
Comment (messense) [inline prerender/app.py]: Might be better to return `Optional[float]`? when it returns `None` don't set these headers.

Comment (Nyoroon) [reply prerender/app.py]: Maybe if `None` then just return current time? We can't check this if `skip_cache` is set.

Comment (messense) [reply prerender/app.py]: Current time should be fine.

**Review by Nyoroon — COMMENTED** [review]
fixed

**Review by messense — APPROVED** [review]
LGTM

File: prerender/cache/disk.py [hunk 12]
```diff
@@ -26,3 +27,7 @@ def __init__(self) -> None:
     def set(self, key: str, payload: bytes, ttl: int = None, format: str = 'html') -> None:
         compressed = lzma.compress(payload)
         self._cache.set(key + format, compressed, expire=ttl)
+
+    async def modified_since(self, key: str, format: str = 'html') -> Optional[float]:
+        stats = await stat(key + format)
```
Comment (mrluanma) [inline prerender/cache/disk.py:32]: It's not normal file here, the DiskCache class is actually a thin wrapper over [diskcache](https://pypi.python.org/pypi/diskcache).

Comment (messense) [reply prerender/cache/disk.py:32]: Fixed in https://github.com/bosondata/chrome-prerender/commit/7edb6d7f691249ca8138be2398cf6bf587835f7f

**Pull request merged**